devito/types/sparse.py

```diff
@@ -218,25 +368,159 @@ def _dist_alltoall(self, dmap=None):
+    def _dist_subfunc_alltoall(self, subfunc, dmap=None):
         """
         The metadata necessary to perform an ``MPI_Alltoallv`` distributing
         self's SubFunction values across the MPI ranks needing them.
         """
-        raise NotImplementedError
+        dmap = dmap or self._dist_datamap
+        ssparse, rsparse = self._dist_count(dmap=dmap)
+
+        # Per-rank shape of send/recv `coordinates`
+        shape = subfunc.shape[1:]
+        sshape = [(i, *shape) for i in ssparse]
+        rshape = [(i, *shape) for i in rsparse]
+
+        # Per-rank count of send/recv `coordinates`
+        scount = [prod(i) for i in sshape]
+        rcount = [prod(i) for i in rshape]
+
+        # Per-rank displacement of send/recv `coordinates` (it's actually all
+        # contiguous, but the Alltoallv needs this information anyway)
+        sdisp = np.concatenate([[0], np.cumsum(scount)[:-1]])
+        rdisp = np.concatenate([[0], tuple(np.cumsum(rcount))[:-1]])
+
+        # Total shape of send/recv `coordinates`
+        sshape = list(subfunc.shape)
+        sshape[0] = sum(ssparse)
+        rshape = list(subfunc.shape)
+        rshape[0] = sum(rsparse)
+
+        return sshape, scount, sdisp, rshape, rcount, rdisp
 
-    def _dist_scatter(self):
+    def _dist_data_scatter(self, data=None):
         """
         A ``numpy.ndarray`` containing up-to-date data values belonging
         to the calling MPI rank. A data value belongs to a given MPI rank R
         if its coordinates fall within R's local domain.
         """
-        raise NotImplementedError
+        data = data if data is not None else self.data._local
+
+        # If not using MPI, don't waste time
+        if self.distributor.nprocs == 1:
+            return data
+
+        # Compute dist map only once
+        dmap = self._dist_datamap
+        mask = self._dist_scatter_mask(dmap=dmap)
+
+        # Pack sparse data values so that they can be sent out via an Alltoallv
+        data = data[mask]
+        data = np.ascontiguousarray(np.transpose(data, self._dist_reorder_mask))
+
+        # Send out the sparse point values
+        _, scount, sdisp, rshape, rcount, rdisp = self._dist_alltoall(dmap=dmap)
+        scattered = np.empty(shape=rshape, dtype=self.dtype)
+        self.comm.Alltoallv([data, scount, sdisp, self._mpitype],
+                            [scattered, rcount, rdisp, self._mpitype])
+
+        # Unpack data values so that they follow the expected storage layout
+        return np.ascontiguousarray(np.transpose(scattered, self._dist_reorder_mask))
+
+    def _dist_subfunc_scatter(self, subfunc):
+        # If not using MPI, don't waste time
+        if self.distributor.nprocs == 1:
+            return {subfunc: subfunc.data}
+
+        # Compute dist map only once
+        dmap = self._dist_datamap
+        mask = self._dist_scatter_mask(dmap=dmap)
+
+        # Pack (reordered) coordinates so that they can be sent out via an Alltoallv
```

Comment:
instead of coordinates, here you mean "subfunc values" ?

Reply:
es